src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1015,103 +1015,145 @@ private void CheckTypesForDuplicates(Assembly assembly)
-        // Version 1.1.1 of Microsoft.CodeAnalysis doesn't provide a good way to handle contract separately from implementation.
-        // To simplify user experience we always add both of them to references.
-        // 1) It's a legitimate scenario, when user provides a custom referenced assembly that was built against the contract assembly
-        // (i.e. System.Management.Automation), so we need the contract one.
-        // 2) We have to provide implementation assembly explicitly, Roslyn doesn't have a way to figure out implementation by itself.
-        // So we are adding both.
-        private static PortableExecutableReference s_objectImplementationAssemblyReference =
-            MetadataReference.CreateFromFile(typeof(object).GetTypeInfo().Assembly.Location);
-
-        private static PortableExecutableReference s_mscorlibAssemblyReference =
-           MetadataReference.CreateFromFile(Assembly.Load(new AssemblyName("mscorlib")).Location);
+        // These assemblies are always automatically added to ReferencedAssemblies.
+        private static Lazy<PortableExecutableReference[]> s_autoReferencedAssemblies = new Lazy<PortableExecutableReference[]>(InitAutoIncludedRefAssemblies);
+        
+        // A HashSet of assembly names to be ignored if they are specified in '-ReferencedAssemblies'
+        private static Lazy<HashSet<string>> s_refAssemblyNamesToIgnore = new Lazy<HashSet<string>>(InitRefAssemblyNamesToIgnore);
 
-        // This assembly should be System.Runtime.dll
-        private static PortableExecutableReference s_systemRuntimeAssemblyReference =
-            MetadataReference.CreateFromFile(ClrFacade.GetAssemblies(typeof(object).FullName).First().Location);
+        // These assemblies are used, when ReferencedAssemblies parameter is not specified.
+        private static Lazy<PortableExecutableReference[]> s_defaultAssemblies = new Lazy<PortableExecutableReference[]>(InitDefaultRefAssemblies);
 
-        // SecureString is defined in a separate assembly.
-        // This fact is an implementation detail and should not require the user to specify one more assembly,
-        // if they want to use SecureString in Add-Type -TypeDefinition.
-        // So this assembly should be in the default assemblies list to provide the best experience.
-        private static PortableExecutableReference s_secureStringAssemblyReference =
-            MetadataReference.CreateFromFile(typeof(System.Security.SecureString).GetTypeInfo().Assembly.Location);
+        private bool InMemory { get { return String.IsNullOrEmpty(outputAssembly); } }
 
+        /// <summary>
+        /// Initialize the list of reference assemblies that will be used when '-ReferencedAssemblies' is not specified.
+        /// </summary>
+        private static PortableExecutableReference[] InitDefaultRefAssemblies()
+        {
+            var defaultRefAssemblies = new List<PortableExecutableReference>(150);
```

Comment:
Currently, `netcoreapp2.0` comes with 137 reference assemblies, and that's why I choose `150` as the default value. I will add a comment.
By default, the capacity of List is 0, and it increases the size as you add more items. So if the size is roughly known, it's better to specify.